web/src/message_events.js

```diff
@@ -102,6 +102,152 @@ export function update_current_view_for_topic_visibility() {
+    if (
+        // Check if current filter relies on the changed message property.
+        !filter_term_types.includes(property_term_type) &&
+        !filter_term_types.includes(`not-${property_term_type}`)
+    ) {
+        return;
+    }
+
+    // We need the message objects to determine if they match the filter.
+    const messages_to_fetch = [];
+    const messages = [];
+    for (const message_id of message_ids) {
+        const message = message_store.get(message_id);
+        if (message !== undefined) {
+            messages.push(message);
+        } else {
+            if (
+                (filter_term_types.includes(property_term_type) && !property_value) ||
+                (filter_term_types.includes(`not-${property_term_type}`) && property_value)
+            ) {
+                // If the message is not cached, that means it is not present in the message list.
+                // Also, the message is not supposed to be in the message list as per the filter and
+                // it's property value. So, we don't need to fetch the message.
+                return;
+            }
+
+            const first_id = msg_list.first().id;
+            const last_id = msg_list.last().id;
+            const has_found_newest = msg_list_data.fetch_status.has_found_newest();
+            const has_found_oldest = msg_list_data.fetch_status.has_found_oldest();
+
+            if (message_id > first_id && message_id < last_id) {
+                // Need to insert message middle of the list.
+                messages_to_fetch.push(message_id);
+            } else if (message_id < first_id && has_found_oldest) {
+                // Need to insert message at the start of list.
+                messages_to_fetch.push(message_id);
+            } else if (message_id > last_id && has_found_newest) {
+                // Need to insert message at the end of list.
+                messages_to_fetch.push(message_id);
+            }
+        }
+    }
+
+    if (!filter.can_apply_locally()) {
+        channel.get({
+            url: "/json/messages",
+            data: {
+                message_ids: JSON.stringify(message_ids),
+                narrow: JSON.stringify(filter.terms()),
+            },
+            success(data) {
+                const messages_to_add = [];
+                const messages_to_remove = new Set(message_ids);
+                for (const raw_message of data.messages) {
+                    messages_to_remove.delete(raw_message.id);
+                    let message = message_store.get(raw_message.id);
+                    if (!message) {
+                        message = message_helper.process_new_message(raw_message);
+                    }
+                    messages_to_add.push(message);
+                }
+                msg_list_data.remove([...messages_to_remove]);
+                msg_list_data.add_messages(messages_to_add);
+                msg_list?.rerender();
+            },
+        });
```

Comment:
We might want to pass an option that lets us consider discarding these from the cache rather than fetching from the server if it's not the current view? 

What seems dangerous about this implementation is that if your last 100 views were all search views, and some messages get moved, this could trigger 100 requests to the server to fetch whether those messages match all of those searches... 

Though do we even currently keep cached `!can_apply_locally` data lists? Maybe the right answer is to ensure we don't do that, and have a strategically placed `assert`.

Reply:
Okay, discarded the cached data if it is not part of a rendered message list.

I think we can store `!can_apply_locally` datasets and use them to render narrows. See the commit
`message_view: Attempt rendering cached data without contacting server.`.

Reply:
OK -- that's fine as long as we have a way to discard those when we cannot ensure they remain accurate.